Generic assets with basic features

cedric-anne opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

First brick of generic asset management.

orthagh reviewed (commented):
A few things for a 1st checkout on my side:

- state vbisibility is broken (as you may filter which state is available for itemtypes)
- model field is not done (maybe a todo to add in #15756 ), this need a big rework as we have separate model tables for each itemtype. We need to have a design like status.
- using assets_assetdefinitions_id=xxx in asset list url make a little ux lose. I think I would prefer something like assetdefintion=xxxname
- More important and related to previous point: you took the root to not have a comprehensive itemtype.
My first test creating a "Car" definition resulted in an itemtype/class name `Glpi\Asset\Asset1` where I guess the 1 is the id of the definition.
I expected having a `Car` Itemtype/classname slugified/derived from the provided label.
I know when a user change the label it could be difficult (impossible) to change the epxected itemtype in others tables but the current choice is a shame.

orthagh commented:
More on the last point, I expect to have before the beta release a combination of this system with public router and a generic route to get pretty url like `http://glpi/car/1`

orthagh commented:
Like discussed together, let's move model thing to later pr.
But sysname / itemtype is mandatory because it drives the direction we want to take

cedric-anne commented on src/Asset/Capacity/AbstractCapacity.php:
```diff
@@ -0,0 +1,207 @@
+     * @param string $linked_itemtype
+     * @return void
+     */
+    protected function deleteRelationLogs(string $source_itemtype, string $linked_itemtype): void
```
If logs are not deleted when capacity is disabled, then the `History` tab will contain entries that are related to disabled capacities and to deleted data, and I think it can be confusing for end-user. If capacity is re-enabled data would not be restored, but log entries will still be present, and it wuld be even more confusing.

I made a test on my machine on a big DB (35M log entries). Removing all 600k lines related to `Monitor` itemtype using a similar query `` DELETE FROM `glpi_logs` WHERE `itemtype` = 'Monitor' OR `itemtype_link` = 'Monitor'; `` takes about 55 seconds. It is indeed too long, even if it is probably an edge case (it would be weird to disable a capacity that is so massively used that it produces so many log entries).

I added an entry in the TODO list in #15756 to try to find a solution for this.

cedric-anne replied:
> My thoughts: I guess it would not be the end of the world if the logs are not immediately deleted and we wait for the cron to clean them.

I added an entry for this in the TODO list. It will be done later.

cedric-anne commented on src/Asset/Asset.php:
```diff
@@ -0,0 +1,393 @@
+            throw new \RuntimeException('Asset definition is expected to de defined in concrete class.');
+        }
+
+        return static::$definition ?? null;
```
I did not find a good way to do this.

Declaring an `abstract public static function getDefinition(): AssetDefinition` method could be a solution, but `atoum` is not able to handle it. Indeed, it breaks tests declared in `tests/functional/Glpi/Asset/Asset.php` as it will automatically tries to create a mock class (that we do not even ask for) and fails with the following error:
```
Error FATAL ERROR in /var/www/glpi/tests/functional/Glpi/Asset/Asset.php on unknown line, generated by file /var/www/glpi/vendor/atoum/atoum/classes/mock/generator.php(219) : eval()'d code on line 2:
Class mock\Glpi\Asset\Asset contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (Glpi\Asset\Asset::getDefinition)
```

The `?? null` was here to fix the `Access to an undefined static property static(Glpi\Asset\Asset)::$definition.` PHPStan issue.

I changed the way it is done in 8a69821c006fae4be60e78e9ad9ea555faddcd06.

If you find a better way to handle this, feel free to propose a patch.

Anyway, as the only expected place to extend this abstract class is in `AssetDefinitionManager::loadConcreteClass()`, this unconventional way of doing things should not cause too many issues.

cedric-anne commented:
> 1. Icon choice as a free text is painful but I suppose it is temporary and a proper icon field will be added later (and it should use tabler-icons instead of font-awesome).

This is in TODO list in #15756:
![image](https://github.com/glpi-project/glpi/assets/33253653/dd4ca2ab-7fba-4fb8-974d-a358760fa3d8)

> 2. When enabling an asset type for your current profile, you can't see it until you reset your session which is a bit annoying and may lead some user to think the feature doesn't work

Fixed by commit 3dcbc445bd141101f60fb07f12cf88a98edddbbd . This is not strictly related to this PR and could probably be reused in 10.0/bugfixes branch for #16031.

cedric-anne commented:
I have some minor changes to do before being ready to merge, but current work could be reviewed. I did not handle schema changes in a migration, so you should drop glpi_assets_* tables and force migration to recreate them.

AdrienClairembault reviewed (changes requested):
Trash bin is not working due to some JS error : 

![image](https://github.com/glpi-project/glpi/assets/42734840/d7abf5c1-23d3-41df-90d0-d15c603ed747)

I disabled notes on my assets and the relevant logs were not deleted (OK for documents et infocom tho):

![image](https://github.com/glpi-project/glpi/assets/42734840/f8fce4ec-3dd1-4137-b917-a5299e468fa9)

Asset definition name is not used as its title:

![image](https://github.com/glpi-project/glpi/assets/42734840/52c4beee-9159-40a1-9edf-385d17a7e998)

Also editing an asset definition in a dangerous operation and we should probably warn the user to confirm when disabling a capacity ("X data will be deleted, are you sure ?").
Can be added to TODO list and done in a later PR, just don't forget about it :)

AdrienClairembault commented:
I've also seen that disabling an asset do not remove it from GLPI until you reset your session.

Its not really related directly to these change as its more more an issue with our session rights that are tied to the current session and never really reloaded.
Its still annoying because I would expect as an administrator that disabling an asset would impact immediately every users.

It kinda get back to the PR where we were handling plugin rights, but I wonder if this kind of thing could be fixed once and for all by adding some kind of rights checksum into the session.
We could check that checksum during the framework initialization and trigger a full rights re-computation if it does not match what we expect.

orthagh commented:
Few small things:
- at creation, I guess a new asset can have all profiles enabled
- about profiles, in fact, we may skip non standard interfaces. They currently don't have access to asset.
- may be added in todo, asset definitions doesn't have their searchoptions (active, comments, etc)
- clearly in todo, we should have the reverse page in profiles form (list the custom assets)
- a question about comprehensive itemtype, and the display of a particular asset (`/front/asset/asset.form.php?id=1`), can we recall the itemtype as a get parameter ? No strictly usefull, but it mays help to do the generic router later.

Remarks about @AdrienClairembault review:

> Asset definition name is not used as its title:

Note, we could use the system name here, but final versions should use translations (planned in todo)

> I disabled notes on my assets and the relevant logs were not deleted (OK for documents et infocom tho):

Do we really want to remove the logs ? I'm not sure any capacities change should impact logs. Did we discussed about that ? I vaguely remember.

> I've also seen that disabling an asset do not remove it from GLPI until you reset your session.

unset `$_SESSION['glpimenu']` after is_enable change should be easy and will regenerate the menu.

AdrienClairembault commented:
> unset $_SESSION['glpimenu'] after is_enable change should be easy and will regenerate the menu.

It will only impact the current user.

Also this problem often come up on many features, thus my suggestion than we could work to fix this globally rather than doing micro fixes here and there.

A simple solution could be to store the date of the last rights changes in the config table and in each user session.
Then if a user load a page and his "date of last rights updates" does not match the one in GLPI's config, we force reload his menu + rights.

cedric-anne commented:
> Trash bin is not working due to some JS error

Fixed.

> I disabled notes on my assets and the relevant logs were not deleted (OK for documents et infocom tho)

Fixed. Even if the way to handle logs may change in the future, all capacities are now handle the same way.

> Asset definition name is not used as its title:

Fixed. It uses the `getTranslatedName()` method, so this will not be forgotten when we would implement the name translations.

> 
> Also editing an asset definition in a dangerous operation and we should probably warn the user to confirm when disabling a capacity ("X data will be deleted, are you sure ?"). Can be added to TODO list and done in a later PR, just don't forget about it :)

Yes, we clearly need to work on a clear UX on this point. Probably a different color for switches that have their values changed could help, and a confirmation message is mandatory.

cedric-anne commented:
> * at creation, I guess a new asset can have all profiles enabled

As discussed IRL, we will keep empty rights at creation, but when no rights are enabled, we will add a message to indicate that profiles must be configured.

> * about profiles, in fact, we may skip non standard interfaces. They currently don't have access to asset.

Done.

> * may be added in todo, asset definitions doesn't have their searchoptions (active, comments, etc)

I added missing search options.

> * clearly in todo, we should have the reverse page in profiles form (list the custom assets)

Added in TODO list.

> * a question about comprehensive itemtype, and the display of a particular asset (`/front/asset/asset.form.php?id=1`), can we recall the itemtype as a get parameter ? No strictly usefull, but it mays help to do the generic router later.

I put back the class in URL parameters.

cedric-anne commented:
I reviewed the profiles rights management part as storing directly in `glpi_profilerights` was a problem. Indeed, when an asset definition was set to inactive, rights were still present, and users were still able to access assets using the corresponding URL.

Now, profiles rights matrix is stored in a `profiles` JSON field. Corresponding rights are pushed into `glpi_profilerights` table when asset definition is set to active, and removed from there when asset definition is set to inactive.

@orthagh @AdrienClairembault A new review could be done. I added a migration for the new field, so it is not necessary to drop the `glpi_assets_*` tables to test this new version.

AdrienClairembault reviewed (changes requested):
Meta criteria doesn't seems to work on generic assets:

![image](https://github.com/glpi-project/glpi/assets/42734840/2ffd2130-bf86-4ec7-9edc-c1f49dd75283)

Only issue I could find, everything else seems OK for me 👍

orthagh commented:
> Meta criteria doesn't seems to work on generic assets:

Not related to this pr (see #16202 and #16212)

orthagh reviewed (approved):
Seems good.
I think we're good to merge, I didn't found neither any bug.
By the way, we're in a state, we can merge this one, and continue in additional pr.

cedric-anne commented:
> Seems good. I think we're good to merge, I didn't found neither any bug. By the way, we're in a state, we can merge this one, and continue in additional pr.

I added some tests today.

Many tests could still be added, but for most of them, they would require to refactor the existing code. For instance, to test the list of items related to a user, the only method that exists for now is `User::showItems()`. To have make it easilly testable, it would require split the logic that builds the list (a method that would return an `iterable`) from the logic that renders the view. This kind of refactoring should be done in separate PRs, to ease the review.

I have done many manual tests on the whole impacted code, so I am pretty confident to merge this right now and to continue the job in subsequent PRs.

cedric-anne commented:
I forgot to say that all TODOs present in the added code corresponds to an entry in #15756. They were left intentionally.
